.github/codeql/queries/assert-pure.ql

```diff
@@ -0,0 +1,29 @@
+/**
+ * @name Assert Pure
+ * @kind problem
+ * @problem.severity error
+ * @id vscode-codeql/assert-pure
+ * @description This query asserts that the modules tested in the pure-tests are "pure".
+ * This means that they don't have any transitive reference to vscode.
+ */
+import javascript
+
+class VSCodeImportDeclaration extends ImportDeclaration {
+  VSCodeImportDeclaration() {
+    this.getImportedPath().getValue() = "vscode"
+  }
+}
+
+ImportDeclaration transitiveImports(Module m) {
+  exists (ImportDeclaration i |
+    (result = i and i = m.getAnImport()) or
+    (result = transitiveImports(m.getAnImport().resolveImportedPath()))
+  )
```

Comment:
As written, the second disjunct is a Cartesian product, because it doesn't talk about `i`. We can pull out the base definition of `i` and use it in both disjuncts.

```
  exists (ImportDeclaration i | i = m.getAnImport() |
    result = i or
    result = transitiveImports(i.resolveImportedPath())
  )
```

As a more general approach, perhaps we can get away with just the following, assuming there is a `Module` entity for `vscode`:
```
where m.getAnImportedModule+().getName() = "vscode"
```

Reply:
The second suggestion does not allow for using `VSCodeImportDeclaration` in the results. So, I think your first suggestion will work better.

Reply:
Good point! It's looking good already now, but for further refinement we can:
- use `Import` everywhere instead of `ImportDeclaration` (so it handles all forms of import, even though in practice we generally stick to ES2015 imports)
- simplify the recursive logic to `where m.getAnImportedModule*().getAnImport() = v`

Reply:
Done.

Reply:
Sorry...tried to be too clever. `Import` is an abstract class so extending it won't work here.
Two options:
- More general but wordy: `class VSCodeImport extends ASTNode { VSCodeImport() { this.(Import).getImportedPath()... }`
- Less general but neater: `class VSCodeImport extends ImportDeclaration { VSCodeImport() { this.getImportedPath()... }`

Reply:
I didn't even know that syntax works.